types/datum.go

```diff
@@ -1930,7 +1930,11 @@ func EqualDatums(sc *stmtctx.StatementContext, a []Datum, b []Datum) (bool, erro
 		return false, nil
 	}
 	for i, ai := range a {
+		collation := ai.Collation()
```

Comment:
I'm a little surprised that `EqualDatum` is only used by `REPLACE INTO`. That's why the change here does not break anything else.

I would suggest moving `EqualDatum` to 'replace.go' and make it a private function so that `EqualDatum` will not be used in an incorrect way.

Reply:
What about the tests of `EqualDatums`? Should we also move them to `write_test.go`?

Reply:
`EqualDatums` is also used in `flamegrapth_test.go`

Reply:
> What about the tests of EqualDatums? Should we also move them to write_test.go?

Yes, that's reasonable if is used by `REPLACE INTO` only.

> EqualDatums is also used in flamegrapth_test.go

I think it can be replaced with `CompareDatum`?